syntax = "proto3";

package hub;

option go_package = "deepapp_golang_grpc_hub/internal/proto";
option java_package = "com.deepapp.hub";

service HubService {
  rpc Connect(stream Message) returns (stream Message);
  rpc UploadFile(stream FileChunk) returns (FileUploadResponse);
  rpc DownloadFile(FileDownloadRequest) returns (stream FileChunk);
}

message Message {
  string id = 1;
  string from = 2;
  string to = 3;
  string channel = 4;
  string content = 5;
  string timestamp = 6;
  MessageType type = 7;
  string action = 8; // For dynamic routing: "register", "request", "response", etc
  map<string, string> metadata = 9; // Flexible metadata
  string request_id = 10; // Unique request identifier for tracking responses
  string file_id = 11; // Reference to uploaded file (instead of base64 in content)
}

// File streaming messages
message FileChunk {
  string file_id = 1;        // Unique file identifier
  string request_id = 2;     // Request this file belongs to
  bytes data = 3;            // Chunk data (max 64KB recommended)
  int64 offset = 4;          // Byte offset in file
  int64 total_size = 5;      // Total file size (-1 if unknown)
  string filename = 6;       // Original filename
  string content_type = 7;   // MIME type
  map<string, string> metadata = 8; // Additional metadata
  bool is_last = 9;          // True for final chunk
}

message FileUploadResponse {
  string file_id = 1;        // Generated file ID
  int64 bytes_received = 2;  // Total bytes received
  string status = 3;         // "success" or "error"
  string error = 4;          // Error message if failed
}

message FileDownloadRequest {
  string file_id = 1;        // File ID to download
  int64 offset = 2;          // Start offset (for resume)
  int64 chunk_size = 3;      // Preferred chunk size
}

enum MessageType {
  DIRECT = 0;
  BROADCAST = 1;
  CHANNEL = 2;
  REGISTER = 3; // Worker registration
  REQUEST = 4;  // Service request
  RESPONSE = 5; // Service response
  WORKER_CALL = 6; // Worker-to-Worker call
}

// Worker registration message
message WorkerRegistration {
  string worker_id = 1;
  string worker_type = 2;
  repeated ServiceCapability capabilities = 3;
  map<string, string> metadata = 4;
}

// Service capability definition
message ServiceCapability {
  string name = 1;
  string description = 2;
  string input_schema = 3; // JSON schema
  string output_schema = 4; // JSON schema
}

message Request {
  RequestType type = 1;
  string data = 2;
}

enum RequestType {
  JSON = 0;
  FILE = 1;
  CONTROL = 2;
}

message Response {
  Status status = 1;
  string data = 2;
}

enum Status {
  OK = 0;
  ERROR = 1;
}